File: Core/Company-Interview/GoldmanSach/GroupAnagrams/GroupAnagrams.java
package GoldmanSachs.GroupAnagrams;

import java.util.*;

    /*
        HashMap
        {
         "[1, 0, 0, 0, 1,...]": ["eat", "tea", "ate"]
        }

        returns: [
                ["eat", "tea","ate"],
                ["tan", "nat"],
                ["bat"]
        ]
    */


public class GroupAnagrams {

    public static void main(String[] args) {
        GroupAnagrams groupAnagInstance = new GroupAnagrams();
        List<List<String>> result = groupAnagInstance
                .groupAnagrams(new String[] { "eat", "tea", "tan", "ate", "nat", "bat" });
        System.out.println(result);
    }

    /**
     * Groups strings that are anagrams of each other.
     * Time Complexity: O(N * K), where N is the number of strings and K is the max length of a string.
     * Space Complexity: O(N * K), to store the grouped strings in the HashMap.
     */
    public List<List<String>> groupAnagrams(String[] strs) {
        // 1. Validation: Handle empty input
        if (strs == null || strs.length == 0) {
            return new ArrayList<>();
        }

        /* * 2. The Strategy:
        * Two words are anagrams if they have the exact same character counts.
        * We use a HashMap where:
        * Key: A unique string representation of the character frequency (e.g., "[1, 0, 2...]")
        * Value: A list of words that match that frequency profile.
        */
        Map<String, List<String>> anagramGroups = new HashMap<>();

        for (String word : strs) {
            // 3. Frequency Counting:
            // Create a count of each letter 'a'-'z'.
            // This is more efficient than sorting for long strings.
            int[] charCount = new int[26];
            for (char c : word.toCharArray()) {
                charCount[c - 'a']++;
            }

            /* * 4. Key Generation:
            * Transform the array into a String to use as a Map key.
            * Arrays.toString() creates a unique "fingerprint" for each anagram group.
            * Example: "eat" and "tea" both become "[1, 0, 0, 0, 1, 0, ..., 1, 0...]"
            */
            String key = Arrays.toString(charCount);

            // Add the word to the appropriate list in the map
            anagramGroups.computeIfAbsent(key, k -> new ArrayList<>()).add(word);
        }

        // 5. Final Output: Return only the grouped lists from the map values
        return new ArrayList<>(anagramGroups.values());
    }
}

File: Core/Algorithms/Neetcode/1-Arrays-Hashing/4-Group-Anagrams/group-anagrams.py
from collections import defaultdict
from typing import List


# Time Complexity:
# Space Complexity:
def groupAnagrams():
    pass


# Time Complexity: O(m * n) — m = number of strings, n = average string length
# Space Complexity:  O(m * n) — storing all strings in the hashmap
def groupAnagramsOptimal(strs: List[str]) -> List[List[str]]:
    # default dict auto-initialises missing keys as empty lists
    # key: tuple of 26 character counts, value: list of anagram strings

    """
    {
        (1,0,1,0...): [],
        ...
    }
    """
    res = defaultdict(list)

    for s in strs:
        # one slot per letter a-z, all starting at zero
        count = [0] * 26
        for c in s:
            # map character to index 0-25:  'a'->0, 'b'->1, ..., 'z'->25
            # then increment that letter's frequency slot
            count[ord(c) - ord("a")] += 1

        # lists are not hashable so convert to tuple to use as dict key
        # anagrams always produce the same frequency tuple → same bucket
        res[tuple(count)].append(s)

    #  return all grouped anagram lists, discarding the keys
    return list(res.values())


print(groupAnagramsOptimal(["eat", "tea", "tan", "ate", "nat", "bat"]))
# print(groupAnagramsOptimal([""]))
# print(groupAnagramsOptimal(["a"]))


# def groupAnagrams(strs: List[str]) -> List[List[str]]:

#     result = {}

#     for s in strs:
#         count = [0] * 26
#         for c in s:
#             count[ord(c) - ord("a")] += 1

#         if tuple(count) not in result:
#             result[tuple(count)] = []
#         result[tuple(count)].append(s)

#     return list(result.values())
